<a href="https://colab.research.google.com/github/martinpius/PG_training/blob/main/nbs.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<p align="justify"><font face="Rage" size=6 color='#d39aed'>Throughout the course, you will engage in interactive lectures,hands-on coding exercises, and group projects using real datasets, all within this notebook. By the end of the training, you will notonly master Python libraries like Pandas, NumPy, Matplotlib, and Scikit-learn but also confidently apply statistical inference,predictive modeling, and data visualization techniques in real-world scenarios.

In [ ]:
!pip install fpdf # Some packages are not available in colab

In [ ]:
# Loading all required Libraries for the entire course:

In [ ]:
import random # For simulation
import pandas as pd # For Data manipulation
import numpy as np # For Computations/Data manipulation
from sklearn.datasets import fetch_california_housing # Sample data
import matplotlib.pyplot as plt # For plotting
import seaborn as sns # For plottings
from sklearn.model_selection import train_test_split # For data spliting
from sklearn.preprocessing import StandardScaler# For Data normalization

# For classical ML
from sklearn.svm import SVR # For Modeling (Support Vector Machine for Regression)
from sklearn.ensemble import RandomForestRegressor # Random Forest regression
from sklearn.tree import DecisionTreeRegressor # DT regression
from sklearn.metrics import mean_squared_error # Evaluation metric
from statsmodels.graphics.gofplots import qqplot # Checking assumpltions LR-models
from scipy import stats # For statistical modeling
import statsmodels.api as sm # For statistical modeling
from sklearn.linear_model import LinearRegression # For LR
from sklearn.metrics import root_mean_squared_error, r2_score # Metrics
from sklearn.linear_model import Ridge, Lasso # Regularized reg
from sklearn.model_selection import cross_val_score #For robust eval
# For time series Analysis we use the following packages
from statsmodels.graphics.tsaplots import plot_acf, plot_pacf # Ploting
from statsmodels.tsa.stattools import adfuller# Statistical testing
from statsmodels.tsa.arima.model import ARIMA # Modeling
from sklearn.metrics import mean_squared_error, mean_absolute_error # TS metrics
from statsmodels.tsa.seasonal import seasonal_decompose# To decompose the TS
from statsmodels.tsa.statespace.sarimax import SARIMAX # Modeling
import json # Saving .json--> rich text fmt
from fpdf import FPDF # report creation
from ipywidgets import interact # To create interactive plots
import plotly.express as px # To create interactive plots


<p align = "left"><font face="Rage" size=6> Day 1: Python Essentials for Statisticians

## 1. Python Syntax Basics

In [ ]:
# Numbers in Python
a = 10 # int
b = 1.2# float


In [ ]:
#=================================

In [ ]:
# Python List:
even = [2,4,6]
odd = list((1,3,5))

In [ ]:
even[2] = 100 # Change the 3rd element to 100
even_ = even * 2 # Concat the same list 2 times

In [ ]:
# res = even - odd  # Not allowed----> change to arr

In [ ]:
even.append([100, 200, 300]) # Append another list to existing one

In [ ]:
# Python tupple: Immutable
even_tuple = (6, 8, 10)

In [ ]:
even_tuple[0] = 100 # change first ellent to 100: immutable

In [ ]:
even_tuple.append(200)# Not allowed

# And many more..

In [ ]:
# Python dictionary
even_dict = {"a": 20, "b": 40, "c": 60}

print(list(even_dict.values())) # print all values
print(list(even_dict.keys())) # print all keys of the dictionary
# Iterate over a dict
for k, v in even_dict.items():
    print(f"key: {k}, value: {v}")
# More can be done...

In [ ]:
# Python Set: {Uni}
set1 = set((1,2,3,4,5,5,6,7,8,8,8,9))
set2 = {1,2,3,4,5,5,5,6,7,8,8,8,9}
print(f"set1: {set1}, set2: {set2}")
# More operations are available such as .add, .remove, ..etc

In [ ]:
# Type conversions for data cleaning
#==============================================================================
raw_value = "34.7"  # Common scenario: string from CSV files
clean_float = float(raw_value) # casting
print(f"Type: {type(clean_float)}, Value: {clean_float}")


Type: <class 'float'>, Value: 34.7


In [ ]:
# Conditional statements in Python:
#===============================================================================
# If statement
x = 10
if x > 5:
    print("x is greater than 5")

# If else statement
x = 3
if x > 5:
    print("x is greater than 5")
else:
    print("x is 5 or less")

# Extension of if else
x = 5
if x > 5:
    print("x is greater than 5")
elif x == 5:
    print("x is exactly 5")
else:
    print("x is less than 5")

age = 25
income = 40000

if age > 18 and income > 30000:
    print("Eligible")

if age < 18 or income < 10000:
    print("Not eligible")

if not (age < 18):
    print("Adult")

# Many tricks are available....

In [ ]:
# Boolean logic for hypothesis testing
effect_present = True
sample_size = 61
sample_adequacy = (sample_size >= 30)
decision = effect_present and sample_adequacy
print(f"Conclusion valid? {decision}")

Conclusion valid? True


In [ ]:
# Application of conditionals for Data analysis

In [ ]:
# Elif ladder for multiple hypothesis outcomes
p_value = 0.07
alpha = 0.05

if p_value < 0.01:
    print("Strong evidence against H₀")
elif p_value < alpha:
    print("Reject H₀")
else:
    print("Fail to reject H₀")


Fail to reject H₀


In [ ]:
# Loops in Python

In [ ]:
# While loop & application for convergence checks (Simple simulation)
iteration = 1
current_error = 15.7
while current_error > 0.5:
    print(f"Iteration {iteration}: Error={current_error}")
    current_error *= 0.8  # Simulating error reduction
    iteration += 1

# More on while & for loops can be added...

Iteration 1: Error=15.7
Iteration 2: Error=12.56
Iteration 3: Error=10.048000000000002
Iteration 4: Error=8.038400000000001
Iteration 5: Error=6.430720000000001
Iteration 6: Error=5.144576000000001
Iteration 7: Error=4.115660800000001
Iteration 8: Error=3.2925286400000005
Iteration 9: Error=2.6340229120000007
Iteration 10: Error=2.1072183296000007
Iteration 11: Error=1.6857746636800006
Iteration 12: Error=1.3486197309440007
Iteration 13: Error=1.0788957847552005
Iteration 14: Error=0.8631166278041604
Iteration 15: Error=0.6904933022433284
Iteration 16: Error=0.5523946417946627


In [ ]:
# Python List & Application in data analysis: Calculate confidence intervals
data = [72, 68, 75, 81, 69, 65, 78, 82, 70, 77]
n = len(data)
mean = sum(data)/n
sorted_data = sorted(data)
lower_ci = sorted_data[int(0.025*n)]
upper_ci = sorted_data[int(0.975*n)]

print(f"Bootstrap 95% CI: [{lower_ci}, {upper_ci}]")
# More operations using Python list can be added

Bootstrap 95% CI: [65, 82]


In [ ]:
# Python tuple and application in storing statistical test parameters

t_test_config = (
    "two-sample",              # test type
    "unequal_var",             # variance assumption
    {"alpha": 0.05, "alternative": "two-sided"}  # test parameters
)

print(f"Performing {t_test_config[0]} t-test with {t_test_config[2]['alternative']} alternative and alpha={t_test_config[2]['alpha']}")
# More operations on tuple can be added

Performing two-sample t-test with two-sided alternative and alpha=0.05


In [ ]:
# Python dictionary and example of a data structure
trial_data = {
    "treatment_group": {
        "n": 150,
        "mean_improvement": 12.7,
        "std_dev": 4.3
    },
    "control_group": {
        "n": 145,
        "mean_improvement": 9.1,
        "std_dev": 5.2
    }
}

# Accessing nested data
effect_size = (trial_data["treatment_group"]["mean_improvement"]
               - trial_data["control_group"]["mean_improvement"])
print(f"Absolute treatment effect: {effect_size:.1f} points")

# Other dictionary operations can be explored

Absolute treatment effect: 3.6 points


In [ ]:
# Python sets and an example on validate survey response categories
allowed_responses = {"strongly_agree", "agree", "neutral", "disagree", "strongly_disagree"}
actual_responses = {"agree", "neutral", "undecided", "agree"}

invalid_responses = actual_responses - allowed_responses
print(f"Invalid responses to filter: {invalid_responses}")

# Other common sets operations can be added

Invalid responses to filter: {'undecided'}


In [ ]:
#Comprehension in Python: An example to create covariance
#====================================================================
# We start with the easier cases such
# -List comprehension, dictionary comprehesion, etc.
#====================================================================

# Nested comprehension


variables = ["X", "Y", "Z"]
cov_matrix = {
    row: {col: 0.5 if row != col else 1.0 for col in variables}
    for row in variables
}

print("Simulated Covariance Matrix:")
for row in cov_matrix:
    print(f"{row}: {cov_matrix[row]}")

# Memory-efficient sum of squares (better for big data)
large_dataset = (x**2 for x in range(1000000))  # Generator, not list
total = sum(large_dataset)
print(f"Sum of squares for 1M elements: {total:,}")

# More examples on Comprehensions can be added

Simulated Covariance Matrix:
X: {'X': 1.0, 'Y': 0.5, 'Z': 0.5}
Y: {'X': 0.5, 'Y': 1.0, 'Z': 0.5}
Z: {'X': 0.5, 'Y': 0.5, 'Z': 1.0}
Sum of squares for 1M elements: 333,332,833,333,500,000


In [ ]:

# Simulate household survey data
def simulate_subjects(subject_id):
    return {
        "id": f"PT-{subject_id}",
        "age": random.randint(18, 90),
        "residense": random.choice(["rural", "urban"]),
        "consuption": f"{random.gauss(100, 15):.4f}",
        "income": f"{random.gauss(215, 10):.4f}" if random.random() > 0.3 else None  # Missing data simulation
    }

# Generate 1000 respondents
study_population = [simulate_subjects(i) for i in range(1000)]

# Show first 3 patients
for subject in study_population[:3]:
    print(subject)

{'id': 'PT-0', 'age': 73, 'residense': 'rural', 'consuption': '112.1441', 'income': None}
{'id': 'PT-1', 'age': 37, 'residense': 'urban', 'consuption': '112.8547', 'income': None}
{'id': 'PT-2', 'age': 52, 'residense': 'rural', 'consuption': '115.3289', 'income': '218.9133'}


In [ ]:
# Error handling: Safely handle missing data
def safe_average(values):
    try:
        clean = [float(x) for x in values if x is not None]
        return sum(clean)/len(clean)
    except ZeroDivisionError:
        print("Warning: All values missing!")
        return None

# Test with real-world missing data
income_results = [p["income"] for p in study_population]
print(f"Average income result: {safe_average(income_results):.1f}")

Average income result: 215.2


In [ ]:
# Exercises in Day 1:
# To be provided based on
# Data types manipulations, conditionals, simulations, errors handling, etc


In [ ]:
# End of Day 1

In [ ]:
#===============================================================================

[link text](https://)# Day 2: Data Management with Pandas and Numpy...

In [ ]:
#===============================================================================

In [ ]:
# Working with Pandas, and Numpy libraries for Data manipulation in Python

pd.set_option('future.no_silent_downcasting', True) # Silence some warnings

In [ ]:
# Understand Pandas Series:--TODO
#->create toy pd.Series and operations

In [ ]:
# Understand Pandas DataFrames:--TODO
#->create toy pd.DataFrame and operations

In [ ]:
# =================================
# Operations on Real Data
#----------------------------------

In [ ]:
# Load dataset from scikit-learn [To demonstrate the rest of Day 2 we use real life dataset]
california = fetch_california_housing()
df = pd.DataFrame(california.data, columns=california.feature_names)

# Add target variable (house value)
df['MedHouseVal'] = california.target * 100000  # Convert to dollar values

print("Dataset shape:", df.shape)
df.head(3)

Dataset shape: (20640, 9)


,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,MedHouseVal
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,452600.0
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,358500.0
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,352100.0


In [ ]:
# Data inspections and cleaning
# 1. Rename columns for clarity
df = df.rename(columns={
    'MedInc': 'MedianIncome',
    'AveRooms': 'AvgRooms',
    'AveBedrms': 'AvgBedrooms',
    'AveOccup': 'AvgOccupancy'
})

# 2. Check metadata
print("Columns:", df.columns.tolist())
print("\nData types:\n", df.dtypes)
print("\nMissing values:\n", df.isnull().sum())

# Add synthetic missing values for demonstration
df.loc[100:150, 'Population'] = np.nan

In [ ]:
print("\nMissing values:\n", df.isnull().sum())

In [ ]:
# Handling Missing data in Python:
# Impute missing Population with median
pop_median = df['Population'].median()
df['Population'] = df['Population'].fillna(pop_median)

# Verify imputation
print("\nMissing values after treatment:\n", df.isnull().sum())

# Other imputations like mean, mode etc

In [ ]:
display(df.head(3))

In [ ]:
# Filtering and transformation
# Create categorical income groups
df['IncomeCategory'] = pd.cut(df['MedianIncome'],
                              bins=[0, 3, 6, 9, 15],
                              labels=['Low', 'Medium', 'High', 'Very High'])


In [ ]:
display(df.head(3))

In [ ]:

# Filter coastal properties
coastal_mask = (df['Latitude'] >= 34) & (df['Latitude'] <= 38) & \
               (df['Longitude'] >= -124) & (df['Longitude'] <= -118)
coastal_housing = df[coastal_mask].copy()


In [ ]:
print(f"Found {len(coastal_housing)} coastal properties")

Found 10136 coastal properties


In [ ]:
# Group analysis by income category
income_stats = df.groupby('IncomeCategory').agg(
    AvgHouseValue=('MedHouseVal', 'mean'),
    MaxBedrooms=('AvgBedrooms', 'max'),
    TotalPopulation=('Population', 'sum')
).reset_index()


<ipython-input-39-0076eaa2c5d4>:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  income_stats = df.groupby('IncomeCategory').agg(


In [ ]:
print("\nIncome group statistics:")
income_stats # You can save if you want

In [ ]:
# Calculate price per room [Feature engineering]
coastal_housing['PricePerRoom'] = coastal_housing['MedHouseVal'] / coastal_housing['AvgRooms']

In [ ]:
display(coastal_housing.head(3))

In [ ]:
# Splitting and Merging Datasets in Pandas
# Split into two datasets
geo_data = df[['Latitude', 'Longitude', 'MedHouseVal']]
economic_data = df.drop(['Latitude', 'Longitude'], axis=1)


In [ ]:
# Merge on index
merged_df = pd.merge(left=geo_data,
                    right=economic_data,
                    left_index=True,
                    right_index=True)

In [ ]:

print("Merged dataset columns:", merged_df.columns.tolist()) # Save the merged data if needed

Merged dataset columns: ['Latitude', 'Longitude', 'MedHouseVal_x', 'MedianIncome', 'HouseAge', 'AvgRooms', 'AvgBedrooms', 'Population', 'AvgOccupancy', 'MedHouseVal_y', 'IncomeCategory']


In [ ]:
# Feature engineering continue..
# Create interaction term: Income-to-Occupancy Ratio
df['IncomeOccupancyRatio'] = df['MedianIncome'] / df['AvgOccupancy']

# Binarize house age
df['OldHouse'] = np.where(df['HouseAge'] > 50, 1, 0)

# Create spatial clusters
df['LatCluster'] = pd.cut(df['Latitude'], bins=5, labels=False)
df['LonCluster'] = pd.cut(df['Longitude'], bins=5, labels=False)

In [ ]:
print("Final dataset info:")
print(f"- Rows: {len(df)}")
print(f"- Columns: {len(df.columns)}")
print("- Sample new features:\n", df[['IncomeOccupancyRatio', 'OldHouse', 'LatCluster']].head(2))

In [ ]:
display(df.head(3))

In [ ]:
# Export and save cleaned dataset for future use
# Save cleaned data for further uses

df.to_csv('california_housing_cleaned.csv', index=False)

# Export income statistics
income_stats.to_excel('income_group_analysis.xlsx', index=False)

In [ ]:
# Day 2 Exercises

In [ ]:
# End of Day 2

In [ ]:
#===============================================================================

# Day 3: EDA

In [ ]:
#===============================================================================

In [ ]:
 # Loading the  Cleaned dataset (The one we saved above)

df = pd.read_csv('california_housing_cleaned.csv')
print("Dataset shape:", df.shape)
df.head(3)

In [ ]:
# Basic descriptive statistics
print("Global Statistics:")
print(df[['MedHouseVal', 'MedianIncome', 'HouseAge']].describe())


In [ ]:
# Skewness/Kurtosis Analysis
print("\nDistribution Shape:")
print(df[['MedHouseVal', 'MedianIncome']].agg(['skew', 'kurtosis']))



In [ ]:
# Correlation Matrix
corr_matrix = df.corr(numeric_only=True)
print("\nTop Correlations with House Value:")
print(corr_matrix['MedHouseVal'].sort_values(ascending=False)[:5])

In [ ]:
# Grouped Analysis
income_group = df.groupby('IncomeCategory').agg(
    AvgHouseValue=('MedHouseVal', 'mean'),
    MedianRooms=('AvgRooms', 'median'),
    OldHousePercentage=('OldHouse', 'mean')
).reset_index()


In [ ]:
print("Income Group Insights:")
income_group.style.format({
    'AvgHouseValue': '${:,.0f}',
    'OldHousePercentage': '{:.1%}'
})

In [ ]:
# Analysis by Geographic Clusters

In [ ]:
geo_analysis = df.groupby(['LatCluster', 'LonCluster']).agg(
    TotalHouses=('MedHouseVal', 'count'),
    PricePerSqMile=('MedHouseVal', lambda x: x.mean()/1000)
).reset_index()

plt.figure(figsize=(10,6))
sns.heatmap(geo_analysis.pivot(index = 'LatCluster', columns = 'LonCluster', values = 'PricePerSqMile'),
            annot=True, cmap='RdYlGn')
plt.title('House Value Density by Geographic Cluster ($K per sq mile)');

In [ ]:
# Distribution Analysis
plt.figure(figsize=(18,6))

# House Value Distribution
plt.subplot(1,2,1)
sns.histplot(df['MedHouseVal'], kde=True, bins=30)
plt.title('House Value Distribution')

# Income-Value Relationship
plt.subplot(1,2,2)
sns.scatterplot(x='MedianIncome', y='MedHouseVal', data=df, alpha=0.3)
plt.title('Income vs House Value')

plt.tight_layout();

Insights:

House values show positive skew (most homes < $300K)

Positive correlation (r=0.69) between income and house value --> Good predictor

In [ ]:
# Anomaly Detection
# Identify price outliers using IQR
Q1 = df['MedHouseVal'].quantile(0.25)
Q3 = df['MedHouseVal'].quantile(0.75)
IQR = Q3 - Q1

outliers = df[(df['MedHouseVal'] < Q1 - 1.5*IQR) |
             (df['MedHouseVal'] > Q3 + 1.5*IQR)]

print(f"Found {len(outliers)} price outliers ({len(outliers)/len(df):.1%} of data)")
print("Outlier characteristics:")
print(outliers[['MedianIncome', 'HouseAge', 'AvgRooms']].describe())

In [ ]:
# Analyse spatial patterns
plt.figure(figsize=(10,6))
sns.scatterplot(x='Longitude', y='Latitude', hue='MedHouseVal',
                data=df, palette='viridis', alpha=0.5)
plt.title('Geographic Price Distribution');

Key Findings:

Coastal regions show higher property values

Bay Area (Lat ~37-38, Lon ~-122) has concentrated high-value homes

#Day 3 Exercises

In [ ]:
# Categorical comparizons:{As questions to the learners (We may choose to hide the cell to let them practice first)}

In [ ]:
# plt.figure(figsize=(10,5))

# # House Age Impact
# sns.boxplot(x='OldHouse', y='MedHouseVal', data=df)
# plt.title('House Value by Age Category (Old = >50 years)')
# plt.ylim(0, 500000);

Interpretation:

Older houses (50+ years) show marginally lower median values

Wider IQR for newer houses indicates greater value variability

In [ ]:
# Pair Plot for Key Features to demonstrate MVA
# sns.pairplot(df[['MedHouseVal', 'MedianIncome', 'HouseAge', 'AvgRooms']],
#              plot_kws={'alpha':0.2})
# plt.suptitle('Multivariate Relationships', y=1.02);

In [ ]:
# The correlation hitmap to identify multicolinearity
# plt.figure(figsize=(10,6))
# mask = np.triu(np.ones_like(corr_matrix, dtype=bool))
# sns.heatmap(corr_matrix, mask=mask, annot=True, cmap='coolwarm')
# plt.title('Feature Correlation Matrix');

In [ ]:
#===============================================================================

In [ ]:
#===============================================================================

# Day 4: Statistical Inference - California Housing Dataset
## 1. Hypothesis Testing
### 1.1 t-Test: For Coastal houses price vs Inland houses Price

In [ ]:
# Re-load the dataset from yesterday:

In [ ]:
df = pd.read_csv("/content/california_housing_cleaned.csv")

In [ ]:
display(df.head(4))

In [ ]:
#===============================================================================

In [ ]:
# Demonstrate the use of using Scipy to carry out statistical analyses


In [ ]:
coastal = df[df['LatCluster'].isin([2,3])]['MedHouseVal']  # Coastal clusters
inland = df[df['LatCluster'].isin([0,4])]['MedHouseVal']   # Inland clusters

# Perform two-sample t-test
t_stat, p_val = stats.ttest_ind(coastal, inland, equal_var=False)
print(f"T-statistic: {t_stat:.2f}, p-value: {p_val:.4f}")

if p_val < 0.05:
    print("Reject H₀: Significant price difference between coastal and inland properties")
else:
    print("No significant difference detected")

T-statistic: -9.66, p-value: 0.0000
Reject H₀: Significant price difference between coastal and inland properties


In [ ]:
# Prepare data groups---> For oneway ANOVA
income_groups = [
    df[df['IncomeCategory'] == cat]['MedHouseVal']
    for cat in ['Low', 'Medium', 'High', 'Very High']
]


In [ ]:
# One-way ANOVA
f_stat, p_val = stats.f_oneway(*income_groups)
print(f"F-statistic: {f_stat:.1f}, p-value: {p_val:.4f}")

F-statistic: 4843.7, p-value: 0.0000


In [ ]:
# 2. Regression Analysis
# 2.1 OLS Regression with StatsModels

In [ ]:
# Prepare the data for MLR

X = df[['MedianIncome', 'AvgRooms', 'HouseAge', 'Latitude']]
X = sm.add_constant(X)  # Adds intercept term
y = df['MedHouseVal'] #Price of the house

# Fit model
model = sm.OLS(y, X).fit()
print(model.summary())

In [ ]:
# Confidence intervals for the regression coeff
# Get 95% CI for coefficients
conf_int = model.conf_int(alpha=0.05)
conf_int.columns = ['2.5%', '97.5%'] # Two tailed test [L=0.05/2, R = 1-]
print("\nConfidence Intervals:")
print(conf_int)

In [ ]:
# Checking the Assumptions for LR

#---> Normality assumption
residuals = model.resid # Fetch the residuals
qqplot(residuals, line='45', fit=True) # Plot the residuals --> q-q plot
plt.title('Q-Q Plot of Residuals')
# Try a transformation of the target variable (e.g., log, square root) if not normal


In [ ]:
#---> Homoscedasticity--# Must be random above and below
plt.scatter(model.fittedvalues, residuals)
plt.axhline(y=0, color='r', linestyle='--')
plt.title('Residuals vs Fitted Values')
#Apply log or square-root transformation to the dependent variable.
#Use weighted least squares (WLS).[When Violeted]

In [ ]:
#===============================================================================

In [ ]:
## Exercise Day 4:
# Calculate confidence interval for Mean Coastal Price
# mean_coastal = coastal.mean()
# sem_coastal = coastal.sem()  # Standard error

# ci_low, ci_high = stats.t.interval(
#     confidence=0.95,
#     df=len(coastal)-1,
#     loc=mean_coastal,
#     scale=sem_coastal
# )

# print(f"Coastal Price 95% CI: (${ci_low:,.0f}, ${ci_high:,.0f})")

In [ ]:
# More to be added interactively!

In [ ]:
#===============================================================================

# Day 5: Predictive Modeling with Scikit-learn - California Housing

In [ ]:
# Loading a dataset from the URL
url = "https://raw.githubusercontent.com/martinpius/PG_training/refs/heads/main/boston_housing.csv"
df = pd.read_csv(url)

In [ ]:
display(df.head(3))

,CRIM,ZN,INDUS,CHAS,NOX,RM,AGE,DIS,RAD,TAX,PTRATIO,B,LSTAT,MEDV
0,0.00632,18.0,2.31,0.0,0.538,6.575,65.2,4.0900,1.0,296.0,15.3,396.90,4.98,24.0
1,0.02731,0.0,7.07,0.0,0.469,6.421,78.9,4.9671,2.0,242.0,17.8,396.90,9.14,21.6
2,0.02729,0.0,7.07,0.0,0.469,7.185,61.1,4.9671,2.0,242.0,17.8,392.83,4.03,34.7


columns: #cols: List = [col for col in columns if col not in ['INDUS','PTRATIO', 'NOX', "RM", "TAX"]]

Variables descriptions:
columns = [
"CRIM", # Per capita crime rate by town
"ZN", # Proportion of residential land zoned for lots over 25,000 sq. ft.
"INDUS", # Proportion of non-retail business acres per town
"CHAS", # Charles River dummy variable (1 if tract bounds river; 0 otherwise)
"NOX", # Nitric oxide concentration (parts per 10 million)
"RM", # Average number of rooms per dwelling
"AGE", # Proportion of owner-occupied units built before 1940
"DIS", # Weighted distances to five Boston employment centers
"RAD", # Index of accessibility to radial highways
"TAX", # Full-value property tax rate per $10,000
"PTRATIO", # Pupil-teacher ratio by town
"B", # 1000(Bk - 0.63)^2 where Bk is the proportion of Black individuals by town
"LSTAT" # Percentage of lower status population

In [ ]:
# Plot the correlation hitmap
plt.figure(figsize = (8, 6))
sns.heatmap(data = df.iloc[:,:-1].corr(), cmap= "coolwarm",
            fmt = ".2f", annot = True)
plt.suptitle("Correlation heatmap", y = 1.02 )

In [ ]:
# Relationships [predictors and target]

plt.figure(figsize = (8, 6))
sns.scatterplot(data = df, x = "CRIM", y = "MEDV", color = "red")
plt.title("ABC")
plt.xlabel("CRIM")
plt.ylabel("Price")
plt.show()

In [ ]:
columns = list(df.columns)
indcols = columns[:-1]


In [ ]:
X = df[indcols]

In [ ]:
y = df["MEDV"]

In [ ]:
Xtrain, Xtest, ytrain, ytest = train_test_split(X, y, random_state = 231, test_size = 0.3)

In [ ]:
reg_model = LinearRegression()

reg_fit = reg_model.fit(X = Xtrain, y = ytrain)
preds = reg_fit.predict(Xtest)



In [ ]:

Rsquare = r2_score(ytest, preds)

In [ ]:

reg_fit.coef_

array([-1.30528035e-01,  4.20375791e-02,  5.31721294e-02,  1.93496035e+00,
       -1.52588101e+01,  3.89628068e+00,  2.93054114e-03, -1.33052203e+00,
        2.79290010e-01, -1.22275357e-02, -9.78726624e-01,  8.45512750e-03,
       -5.63880415e-01])

In [ ]:

mse = mean_squared_error(ytest, preds)

In [ ]:
mse

21.110958843971083

In [ ]:

rmse = root_mean_squared_error(ytest, preds)

In [ ]:

rmse

In [ ]:
# Use regularized reg for improvement (if any)

In [ ]:


# Ridge Regression
ridge = Ridge(alpha= 1.5)
ridge.fit(Xtrain, ytrain)
ridge_rmse = mean_squared_error(ytest, ridge.predict(Xtest))

# Lasso Regression
lasso = Lasso(alpha= 0.0003)
lasso.fit(Xtrain, ytrain)
lasso_rmse = mean_squared_error(ytest, lasso.predict(Xtest))

print(f"Ridge RMSE: ${ridge_rmse:,.0f}")
print(f"Lasso RMSE: ${lasso_rmse:,.0f}")

In [ ]:


# 5-fold cross-validation for Ridge
ridge_scores = cross_val_score(
    Ridge(alpha=1.0),
    Xtrain,
    ytrain,
    scoring='neg_root_mean_squared_error',
    cv=5
)
print(f"Ridge CV RMSE: ${-ridge_scores.mean():,.0f} (±${ridge_scores.std():,.0f})")

In [ ]:
# Model selection/Interpretation

In [ ]:
coeffs = pd.DataFrame({
    'Feature': X.columns,
    'LR': reg_fit.coef_,
    'Ridge': ridge.coef_,
    'Lasso': lasso.coef_
}).set_index('Feature')

print("Feature Coefficients:")
coeffs.style.format("{:,.5f}")

In [ ]:
# ML Intro--> Classical models [DT, RF, & SVM]

In [ ]:

# DT for reg
dt = DecisionTreeRegressor()
dt.fit(Xtrain, ytrain)
y_pred_dt = dt.predict(Xtest)

In [ ]:

# SVM-for reg
svm = SVR()
svm.fit(Xtrain, ytrain)
y_pred_svm = svm.predict(Xtest)

In [ ]:
# RF- for reg
rf = RandomForestRegressor()
rf.fit(Xtrain, ytrain)
y_pred_rf = rf.predict(Xtest)

In [ ]:

# Metrics
print("SVM RMSE:", mean_squared_error(ytest, y_pred_svm))
print("RF RMSE:", mean_squared_error(ytest, y_pred_rf))
print("DT RMSE:", mean_squared_error(ytest, y_pred_dt))

In [ ]:
#===============================================================================

In [ ]:
# Time Series Forecasting: Basics


In [ ]:
#===============================================================================

In [ ]:
# Loading the data from the URL:
url = "https://raw.githubusercontent.com/martinpius/LargeLaN/main/dfm_all.csv"

In [ ]:
# Download and preprocess the data
df = pd.read_csv(url)



In [ ]:
display(df.head(4))

In [ ]:
df["date"] = pd.to_datetime(df["date"]) #Convert the date column into a proper date variable

In [ ]:
df.set_index("date", inplace = True) # indexing the dataframe by the date

In [ ]:
df = df.resample(pd.Timedelta(15, "m"), closed = "right").mean().interpolate() # Data were collected every 15 minutes

In [ ]:
display(df.head(4))

In [ ]:
# Plotting the baseline model:
plt.figure(figsize = (12, 4))
# Plot the baseline model predictions using 2018 data
plt.plot(df.loc["2018-01-01 00:00:00":]['prediction'],
                 color = "gray", label = "prediction")
plt.plot(df.loc["2018-01-01 00:00:00":]['measurement'],
                 color = "fuchsia", label = "real")
plt.title("Performance Evaluation: Baseline model")
plt.xlabel("2018 data")
plt.ylabel("Electricty consuption in (KWhrs)")
plt.legend(loc = "best")
plt.show()

In [ ]:
# EDA

In [ ]:
# Focus on the measurement column
ts = df['measurement']

In [ ]:
# Ensure 15-minute frequency
ts = ts.asfreq('15min')

In [ ]:
ts.head(2)

,measurement
date,
2015-12-31 23:45:00,593.548667
2016-01-01 00:00:00,600.781583


In [ ]:
# Vizualize the raw series:
plt.figure(figsize=(14,5))
plt.plot(ts, color = "lightblue")
plt.title("Electricity consuption per 45 min")
#ts.plot(title='Electricity Demand Time Series', color)
plt.ylabel("Demand (kW)")
plt.grid(False)
plt.show()

In [ ]:
#  Resample and visualize trends [For smoothing--> rectifying]
ts_hourly = ts.resample('h').mean() # Hourly
ts_daily = ts.resample('D').mean() # Daily

fig, axs = plt.subplots(2, 1, figsize=(14,8))
axs[0].plot(ts_hourly, color = "lightgray")
axs[1].plot(ts_daily, color = "lightpink")
axs[0].set_title("Hourly Average Demand")
axs[1].set_title("Daily Average Demand")
plt.tight_layout()
plt.show()

In [ ]:
# Import more helpers' libraries

In [ ]:
# Decompose[trend+seasonal + noise] (optional, only if seasonal_decompose available)
decomposition = seasonal_decompose(ts, model='additive', period=96)  # 96 = 1 day of 15-min data
# Plotting the decomposed components
decomposition.plot()
plt.show()

In [ ]:
# Ensure stationarity: Test using ADF

result = adfuller(ts.dropna())
print(f"ADF Statistic: {result[0]}")
print(f"p-value: {result[1]}")

In [ ]:
#In case of non-stationary data
# Differencing is needed
#ts_diff = ts.diff().dropna()#---> No need , the TS is stationary

In [ ]:
# Determine the order of ARMA (p,d = 0,  q)
fig, axs = plt.subplots(2, 1, figsize=(14, 8))  # Wider and taller
plot_acf(ts, lags=50, ax=axs[0])
axs[0].set_title('Autocorrelation Function (ACF)')
plot_pacf(ts, lags=50, ax=axs[1])
axs[1].set_title('Partial Autocorrelation Function (PACF)')
plt.tight_layout()
plt.show()

In [ ]:
# The order of MA--> ACF, AR---> PACF (Great variations can be seen)

In [ ]:
# Train-test split (forecasting 2 days = 192 steps)
train = ts.loc[:"2017-12-31 23:45:00"]
test = ts.loc["2018-01-01 00:00:00":]

In [ ]:
#  Fit ARIMA model (example: p=2, d=0, q=2; tune this)
model = ARIMA(train, order=(2,0,2))
model_fit = model.fit()
print(model_fit.summary())

In [ ]:
# Forecasting
forecast = model_fit.forecast(steps=192)

In [ ]:
# Evaluation
rmse = np.sqrt(mean_squared_error(test[:192], forecast))
mae = mean_absolute_error(test[:192], forecast)
mape = np.mean(np.abs((test[:192] - forecast) / test)) * 100

In [ ]:
print(f"RMSE: {rmse:.2f}")
print(f"MAE: {mae:.2f}")
print(f"MAPE: {mape:.2f}%")

In [ ]:
# Plot forecast vs actual
plt.figure(figsize=(14,5))
plt.plot(test[:192].index, test[:192], label='Actual')
plt.plot(test[:192].index, forecast, label='Forecast')
plt.title("2-Day Ahead Forecast vs Actual")
plt.legend()
plt.grid(True)
plt.show()

In [ ]:
# Fitting SARIMA



In [ ]:
# Defining  and fitting SARIMA the model [Take a lot of time...try @ home]
# model = SARIMAX(ts,
#                 order=(2, 0, 2),       # non-seasonal p, d, q
#                 seasonal_order=(1, 1, 1, 96),  # seasonal P, D, Q, s
#                 enforce_stationarity=False,
#                 enforce_invertibility=False)
# results = model.fit()
# print(results.summary())

In [ ]:
# # 2 -days ahead forecasting:
# forecast = results.get_forecast(steps=96)
# forecast_ci = forecast.conf_int()


In [ ]:
#===============================================================================

In [ ]:
# Exercises Day 5
# Play around with p,d, q, P, D, Q, s---> Find the best model

In [ ]:
#===============================================================================

In [ ]:
#==============================================================================


# Day 6: Automating Analysis.

In [ ]:
#===============================================================================

In [ ]:
## 1. Reusable Functions for Common Tasks

def load_housing_data(file_path='california_housing_cleaned.csv'):
    """Load and preprocess data with consistent formatting"""
    df = pd.read_csv(file_path)
    #display(df.head(3))
    df['IncomeCategory'] = pd.cut(df['MedianIncome'], bins=[0, 3, 6, 9, 15],
                                  labels=['Low', 'Medium', 'High', 'Very High'])
    return df

In [ ]:
def save_results(results_dict):
    """Save analysis outputs to JSON"""
    with open('final_analysis.json', 'w') as f:
        json.dump(results_dict, f)
    print(f"Results saved to {'final_analysis.json'}")

In [ ]:
DATA_PATH = "/content/california_housing_cleaned.csv"
OUTPUT_DIR = 'content/'

In [ ]:
def run_full_analysis():
    """End-to-end analysis workflow-reusing the above"""
    df = load_housing_data(DATA_PATH)
    df.drop("IncomeCategory", axis = 1, inplace = True)

    # Perform analysis
    summary_stats = df.describe().to_dict()
    corr_matrix = df.corr().to_dict()

    # Save outputs
    save_results({'stats': summary_stats, 'correlations': corr_matrix})

    return summary_stats, corr_matrix

In [ ]:
sumsta, cor = run_full_analysis()

Results saved to final_analysis.json


In [ ]:
# Batch processing automated:
df = pd.read_csv(DATA_PATH)

# Batch process multiple parameter combinations
param_grid = {
    'income_bins': [[0, 3, 6, 9, 15], [0, 5, 10, 15]],
    'age_thresholds': [40, 50, 60]
}

results = []

for bins in param_grid['income_bins']:
    for threshold in param_grid['age_thresholds']:
        # Automated categorization
        df['IncomeGroup'] = pd.cut(df['MedianIncome'], bins=bins)
        df['OldHouse'] = np.where(df['HouseAge'] > threshold, 1, 0)

        # Store configuration and result
        results.append({
            'bins': bins,
            'threshold': threshold,
            'old_houses_pct': df['OldHouse'].mean()
        })

pd.DataFrame(results).to_csv('parameter_sweep_results.csv', index=False)

In [ ]:
# Automate report
# Generate PDF report with results

def create_report(summary_stats, output_file='report.pdf'):
    """Convert analysis results to PDF"""
    pdf = FPDF()
    pdf.add_page()
    pdf.set_font("Arial", size=12)

    pdf.cell(200, 10, txt="California Housing Analysis Report", ln=1, align='C')

    # Add summary table
    pdf.cell(200, 10, txt="Key Statistics:", ln=1)
    for stat, value in summary_stats['MedHouseVal'].items():
        pdf.cell(200, 10, txt=f"{stat}: {value:.2f}", ln=1)

    pdf.output(output_file)

# Usage
stats = df.describe().to_dict()
create_report(stats)

In [ ]:
# Interactive Dashboard

In [ ]:
@interact
def show_metrics(Income_Category=['Low', 'Medium', 'High', 'Very High']):
    subset = df[df['IncomeCategory'] == Income_Category]
    display(subset.describe())

    plt.figure(figsize=(8,4))
    sns.boxplot(x='OldHouse', y='MedHouseVal', data=subset)
    plt.show()

In [ ]:

# Load data from URL (Other version of housing price)
df = pd.read_csv('https://raw.githubusercontent.com/ageron/handson-ml2/master/datasets/housing/housing.csv')



In [ ]:
display(df.head(3))

In [ ]:
# Interactive dashboard
@interact
def housing_dashboard(
    Ocean_Proximity=df['ocean_proximity'].unique(),
    Min_Income=(0.0, 15.0, 0.5),
    Max_House_Age=(0, 100, 5)
):
    filtered = df[
        (df['ocean_proximity'] == Ocean_Proximity) &
        (df['median_income'] >= Min_Income) &
        (df['housing_median_age'] <= Max_House_Age)
    ]

    # Create interactive plot
    fig = px.scatter(
        filtered,
        x="median_income",
        y="median_house_value",
        color="population",
        size="total_rooms",
        hover_name="ocean_proximity"
    )
    fig.show()

    # Show stats
    print(f"Found {len(filtered)} properties")
    display(filtered.describe())

In [ ]:
#===============================================================================

In [ ]:
# Day 6 Exercise

In [ ]:
#===============================================================================

#Day 7: Mini-project, presentations (group presentations), etc + Evaluation + Conclusion